pkg/analyzer/lib/src/generated/gn.dart

```diff
@@ -8,202 +8,126 @@ import 'dart:collection';
 import 'dart:core';
 
 import 'package:analyzer/file_system/file_system.dart';
+import 'package:analyzer/source/package_map_resolver.dart';
 import 'package:analyzer/src/generated/sdk.dart';
 import 'package:analyzer/src/generated/source.dart';
 import 'package:analyzer/src/generated/source_io.dart';
 import 'package:analyzer/src/generated/workspace.dart';
+import 'package:package_config/packages.dart';
+import 'package:package_config/packages_file.dart';
+import 'package:package_config/src/packages_impl.dart';
 import 'package:path/path.dart';
 
 /**
- * Similar to Map#putIfAbsent, except that a value is stored only if it is not
- * null.
- */
-V _putIfNotNull<K, V>(Map<K, V> map, K key, V ifAbsent()) {
-  if (map.containsKey(key)) {
-    return map[key];
-  }
-  V computed = ifAbsent();
-  if (computed != null) {
-    map[key] = computed;
-  }
-  return computed;
-}
-
-/**
- * The [UriResolver] used to resolve `file` URIs in a [GnWorkspace].
+ * Information about a Gn workspace.
  */
-class GnFileUriResolver extends ResourceUriResolver {
+class GnWorkspace extends Workspace {
   /**
-   * The workspace associated with this resolver.
+   * The name of the directory that identifies the root of the workspace.
    */
-  final GnWorkspace workspace;
+  static const String _jiriRootName = '.jiri_root';
 
   /**
-   * Initialize a newly created resolver to be associated with the given
-   * [workspace].
+   * The resource provider used to access the file system.
    */
-  GnFileUriResolver(GnWorkspace workspace)
-      : workspace = workspace,
-        super(workspace.provider);
-
-  @override
-  Source resolveAbsolute(Uri uri, [Uri actualUri]) {
-    if (!ResourceUriResolver.isFileUri(uri)) {
-      return null;
-    }
-    String path = provider.pathContext.fromUri(uri);
-    File file = workspace.findFile(path);
-    if (file != null) {
-      return file.createSource(actualUri ?? uri);
-    }
-    return null;
-  }
-}
+  final ResourceProvider provider;
 
-/**
- * The [UriResolver] used to resolve `package` URIs in a [GnWorkspace].
- */
-class GnPackageUriResolver extends UriResolver {
   /**
-   * The workspace associated with this resolver.
+   * The absolute workspace root path (the directory containing the `.jiri_root`
+   * directory).
    */
-  final GnWorkspace workspace;
+  final String root;
 
   /**
-   * The path context that should be used to manipulate file system paths.
+   * The path to the directory with source locations.
+   *
+   * Each file in this directory is named after a Dart package and contains the
+   * path to the package's sources.
    */
-  final Context pathContext;
+  String _packagesFilePath;
```

Comment:
The comment should be updated. Based on the name I'm guessing this is the path to a file, not a directory.